p2panda-rs/src/document/document.rs

```diff
@@ -0,0 +1,474 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::collections::BTreeMap;
+use std::convert::TryFrom;
+
+use crate::document::{DocumentBuilderError, DocumentError};
+use crate::hash::Hash;
+use crate::identity::Author;
+use crate::instance::Instance;
+use crate::operation::{AsOperation, OperationWithMeta};
+use crate::schema::{Schema, ValidateOperation};
+use incremental_topo::IncrementalTopo;
+
+use crate::Validate;
+
+/// Hard coded cddl string for now
+const DOCUMENT_SCHEMA: &str = "cafe = { (
+    name: { type: \"str\", value: tstr }
+) }";
+
+/// A resolvable data type made up of a collection of causally linked operations.
+#[derive(Debug)]
+pub struct Document {
+    /// The hash id of this document, it is the hash of the entry of this documents root CREATE operation.
+    id: Hash,
+    /// The hash id of the schema operations in this document follow.
+    schema: Schema,
+    /// The author (public key) who published the CREATE message which instantiated this document.
+    author: Author,
+    /// Permissions, derived from KeyGroup relations, which apply to the document as a whole or operation ranges within it.
+    /// TODO: this is a placeholder for now.
+    #[allow(dead_code)]
+    permissions: Option<Vec<Author>>,
```

Comment:
I would remove everything related to permissions for now, better not mention it at all when its not fleshed out yet :+1:

Reply:
Yep, good shout.